RFC/src/RFC-0110_BaseNodes.md

```diff
@@ -138,6 +138,42 @@ Valid transactions are:
 * added to the [mempool];
 * forwarded to peers using the transaction [BroadcastStrategy].
 
+#### Block/Transaction Weight 
+[block-transaction weight]: #blocktransaction-weight "Block/Transaction Weight"
+
+The weight of a transaction / block measured in "grams". Input, output and kernel weights reflect their respective relative 
+storage and computation cost. Transaction fees are typically proportional to a transaction body's total weight, creating 
+incentive to reduce the size of the UTXO set.
+
+|                          	| Bytes    	| Weight (grams) 	| Notes                                                               	|
+|--------------------------	|----------	|----------------	|---------------------------------------------------------------------	|
+| Output:                  	|          	|                	|                                                                     	|
+| - Per output             	| ~777     	| 13             	| Incl. Rangeproof                                                    	|
+| - Asset Registration     	| >= 256   	| 4              	| Set in OutputFeatures, size depends on number of template IDs used 	|
+| - Tari script per opcode 	| variable 	| 1              	|                                                                       |
+| Input:                   	|          	|                	|                                                                     	|
+| - Per input              	| ~196     	| 1              	| Excl. script field                                                  	|
+| Kernel:                  	|          	|                	|                                                                     	|
+| - Per kernel             	| 112      	| 3              	| Excess signature verification                                       	|
+
```

Comment:
For context, the original weightings were calculated as follows

* Closely correlated to the space taken up on the blockchain
* Smallest element (input) has a weight of 1. Everything else scaled to that.
* Max block weight set to have max block sizes of ±1MB

Computation time wasn't taken into account, but this is an interesting idea.

Now clearly, kernels are the smallest contributor to the block, so you might want to set that weight to 1
Inputs are almost 2. With script some script inputs it'll be larger than 2. (You didn't indicate how you got to 196 bytes).

Outputs are now 7. But there is an argument for making this even larger since we know that RP validation is the most expensive part of tx validation currently.

With scripts, we could go flat rate per opcode, or we could go full ethereum and assign a weight for each opcode. a `PushZero` is basically free, both in space and time, while `CheckMultiSig(16,32,...)` is computationally and spatially expensive.

A middle ground is to assign something like 1.5  grams per byte of the serialised script to recover some of the computationally expensive operations (the exact number will require some benchmarking and investigation).

Reply:
That's great thanks, pretty much exactly how I thought about the original weightings and I got the same values [1,2,7] as you given the byte sizes (quickly summed up from the respective structs, pk compressed only) - though I decided not to deviate from current weightings without some discussion.   

I suppose the challenge is how to go about deciding on values for these while being cognisant of relative impacts on total weights, unintentional incentives, complexity, performance etc. My feeling is that TariScript serialisation is likely fast enough to use during verification (or perhaps even use raw bytes before deserialisation), avoids the task of maintaining weightings per OpCode and we can target approx. the same number of NoOp transactions per 1MB block - with larger scripts/assert reg. eating into the max total block weight as appropriate. I suspect the factor per byte will be < 1 since e.g. a single ~32 byte PushPubkey would take the same weight as ~4.5 outputs (or 2.5 if outputs are weighted 13) - or per output etc weightings should be made larger. 

I'll investigate and run the numbers in a spreadsheet and see what I can come up with

Reply:
Right, it's < 1 weight factor. I was implicitly suggesting 1.5 times per byte. So the weight would be  more like `1.5/112 = 0.014`.

Reply:
Using computation is a good metric for the scripts, but we need to look at the weights carefully.  Although a kernel is the smallest unit they don't get pruned. We want to force the user to use as few as possible kernels in his transaction as they stay forever and never get pruned. Rangeproofs are big and computationally expensive, that's why we should weigh them big and they are. We want to spend their inputs, so we weigh them as little as possible. 

I think it's important to keep in mind why we use weights in the blocks. 
1: It's to keep the block to a certain size. 
2: We use the weights to pay fees proportionately. 

But with point 2, we also have to remember we use fees to incentivize users to use the correct behavior. What is the best ideal case: Spend as many inputs as possible, make as few as possible outputs, use as few as possible kernels, and use as fast/small scripts as possible. The fees and in turn the weights should follows this.